oracle/src/main.rs

```diff
@@ -67,3 +80,108 @@ fn main() -> eyre::Result<()> {
+            last_price: "40100.00".to_string(),
+            last_quantity: "0.5".to_string(),
+            best_bid_price: "40095.00".to_string(),
+            best_bid_quantity: "1.0".to_string(),
+            best_ask_price: "40105.00".to_string(),
+            best_ask_quantity: "1.0".to_string(),
+            open_price: "39900.00".to_string(),
+            high_price: "40500.00".to_string(),
+            low_price: "39800.00".to_string(),
+            volume: "1500".to_string(),
+            quote_volume: "60000000".to_string(),
+            open_time: 1698237050000,
+            close_time: 1698323450000,
+            first_trade_id: 1,
+            last_trade_id: 2000,
+            num_trades: 2000,
+        };
+
+        // Wrap the Ticker in DataFeeds::Binance
+        let data_feed = DataFeeds::Binance(ticker);
+
+        // Create a stream that sends a single item and then ends, boxed and pinned
+        Box::pin(stream::once(async { Ok(data_feed) }))
+    }
+
+    #[tokio::test]
+    async fn e2e_oracles() {
+        reth_tracing::init_test_tracing();
+
+        // spawn first instance
+        let (ctx_1, _handle) = test_exex_context().await.unwrap();
+        let data_feed1 = mock_stream();
+        let (oracle_1, network_1) = start(ctx_1, 30303, 30304, data_feed1).await.unwrap();
+        let mut broadcast_stream_1 = BroadcastStream::new(oracle_1.signed_ticks().subscribe());
+        let signer_1 = oracle_1.signer().address();
+        tokio::spawn(oracle_1);
+        let net_1_events = network_1.event_listener();
+
+        // spawn second instance
+        let (ctx_2, _handle) = test_exex_context().await.unwrap();
+        let data_feed2 = mock_stream();
+        let (oracle_2, network_2) = start(ctx_2, 30305, 30306, data_feed2).await.unwrap();
+        let mut broadcast_stream_2 = BroadcastStream::new(oracle_2.signed_ticks().subscribe());
+        let signer_2 = oracle_2.signer().address();
+        tokio::spawn(oracle_2);
+        let net_2_events = network_2.event_listener();
+
+        // expect peers connected
+        let (peer_2, addr_2) = (network_2.peer_id(), network_2.local_addr());
+        network_1.add_peer(*peer_2, addr_2);
+        let expected_peer_2 = wait_for_session(net_1_events).await;
+        assert_eq!(expected_peer_2, *peer_2);
+
+        let (peer_1, addr_1) = (network_1.peer_id(), network_1.local_addr());
+        network_2.add_peer(*peer_1, addr_1);
+        let expected_peer_1 = wait_for_session(net_2_events).await;
+        assert_eq!(expected_peer_1, *peer_1);
+
+        // expect signed data
+        let signed_ticker_1 = broadcast_stream_1.next().await.unwrap().unwrap();
+        assert_eq!(signed_ticker_1.ticker.symbol, "BTCUSDT");
+        assert_eq!(signed_ticker_1.signer, signer_1);
+
+        let signed_ticker_2 = broadcast_stream_2.next().await.unwrap().unwrap();
+        assert_eq!(signed_ticker_2.ticker.symbol, "BTCUSDT");
+        assert_eq!(signed_ticker_2.signer, signer_2);
```

Comment:
so far this tests that both instances received an item from the stream, but it doesn't test if they actually exchange messages as peers after connecting. Can we do that as well?